### PR TITLE
feat: add policy for preventing unencrypted uploads, add new output, add possibility to create tfstate-backend by terraform

### PR DESCRIPTION
**Pull request by SweetOps**
## What
* add policy for preventing unencrypted uploads
* add new output
* add possibility to create tfstate-backend by terraform

### CODE REVIEW

**Comment by SweetOps** [comment]
```hcl
module "terraform_state_backend" {
  source                             = "../terraform-aws-tfstate-backend"
  namespace                          = "zb"
  stage                              = "prod"
  name                               = "terraform"
  attributes                         = ["state"]
  region                             = "eu-central-1"
  prevent_unencrypted_uploads        = "true"
  terraform_backend_config_file_path = "."
}

output "terraform_backend_config" {
  value = "${module.terraform_state_backend.terraform_backend_config}"
}
```
Screenshot 2019-03-18 13 59 38

File: output.tf [hunk 1]
```diff
@@ -27,3 +27,8 @@ output "dynamodb_table_arn" {
   value       = "${element(coalescelist(aws_dynamodb_table.with_server_side_encryption.*.arn, aws_dynamodb_table.without_server_side_encryption.*.arn), 0)}"
   description = "DynamoDB table ARN"
 }
+
+output "terraform_backend_config" {
+  value       = "${data.template_file.terraform_backend_config.rendered}"
+  description = "Redenred terraform backend config file"
```
Comment (aknysh) [inline output.tf]:
```suggestion
  description = "Rendered Terraform backend config file"
```

File: main.tf [hunk 2]
```diff
@@ -18,11 +24,72 @@ module "s3_bucket_label" {
   context = "${module.base_label.context}"
 }
 
+data "aws_iam_policy_document" "prevent_unencrypted_uploads" {
```
Comment (aknysh) [inline main.tf]: maybe make it optional, not all projects require this policy to be applied to the bucket

File: templates/terraform.tf.tpl [hunk 3]
```diff
@@ -0,0 +1,11 @@
+terraform {
+  required_version = ">= 0.11.3"
```
Comment (aknysh) [inline templates/terraform.tf.tpl]: just wondering why do we need to output this template? :) 
why not just output all the values in regular outputs (if we rae missing some in the existing output.tf)?

File: main.tf (hunk 2 above)
Comment (SweetOps) [reply main.tf]: this's resource already optional :) 

https://github.com/cloudposse/terraform-aws-tfstate-backend/pull/17/files#diff-7a370d8342e7203b805911c92454f0f4R2

https://github.com/cloudposse/terraform-aws-tfstate-backend/pull/17/files#diff-7a370d8342e7203b805911c92454f0f4R2

https://github.com/cloudposse/terraform-aws-tfstate-backend/pull/17/files#diff-8e2475e53e84b1e038e9f5d670c3faefR20

File: templates/terraform.tf.tpl (hunk 3 above)
Comment (SweetOps) [reply templates/terraform.tf.tpl]: you know that feel when you refactoring someone else's code and removed backend config by mistake and don't remember in which file it was... of course you can do reset but for what, more easy way it copy-paste from previous `plan/apply` :) 

And also I working with various backends, so for me is hard to remember how to configure them. ` just output all the values in regular outputs` for me it's just list of some values :)

Comment (osterman) [inline templates/terraform.tf.tpl]: I think this version should be parameterized as different environments will have different versions.

File: templates/terraform.tf.tpl [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+  backend "s3" {
+    region         = "${region}"
+    bucket         = "${bucket}"
+    key            = "terraform.tfstate"
```
Comment (osterman) [inline templates/terraform.tf.tpl]: Let's parameterize this as well as we have it parameterized in `geodesic` https://github.com/cloudposse/geodesic/blob/master/rootfs/etc/direnv/rc.d/terraform#L14

File: templates/terraform.tf.tpl (hunk 3 above)
Comment (osterman) [reply templates/terraform.tf.tpl]: We also use `profile` &/|| `role_arn`. I think it's a slippery slope templatizing the config, but it's an interesting notion.

**Comment by osterman** [comment]
@SweetOps pretty cool suggestion!

File: variables.tf [hunk 5]
```diff
@@ -119,3 +119,39 @@ variable "regex_replace_chars" {
   default     = "/[^a-zA-Z0-9-]/"
   description = "Regex to replace chars with empty string in `namespace`, `environment`, `stage` and `name`. By default only hyphens, letters and digits are allowed, all other chars are removed"
 }
+
+variable "prevent_unencrypted_uploads" {
+  type        = "string"
+  default     = "false"
```
Comment (osterman) [inline variables.tf]:
```suggestion
  default     = "true"
```

Let's change default to `true` since that's the best practice.

Comment (SweetOps) [reply variables.tf]: done

**Pull request merged**